[PHX-2088] chore: Make writing into a file more robust

radmiratbrighterai opened the pull request:
Return actual saved path.
Raise exception in case of failure

pkern90 commented on pyproject.toml:
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "redact"
-version = "10.0.1"
+version = "11.0.0"
```
We discussed internally and we are fine with going to `10.1.0` it's technically not 100% correct but should be fine in this case. Retuning a Path instead of None should be a non breaking change. Only the exception is kind of a breaking change.

pkern90 commented on redact/v4/redact_job.py:
```diff
@@ -44,8 +45,10 @@ def download_result(self, ignore_warnings: bool = False) -> JobResult:
-        self.redact.write_output_to_file(
+    def download_result_to_file(
+        self, file: Path, ignore_warnings: bool = False
+    ) -> Optional[Path]:
```
Why is the return value still Optional?

radmiratbrighterai replied:
removed the optional

pkern90 commented on redact/v4/redact_requests.py:
```diff
@@ -201,12 +201,20 @@ def _stream_output_to_file(
-                    os.unlink(temp_file.name)
+                    target_file.unlink(missing_ok=True)
+
+                    raise RuntimeError(f"failed to download the file {file}")
```
Please create a custom Error inside `redact/errors.py` which is more meaning full. Something like:
```
class FileDownloadError(Exception):
    """Raised when a file cannot be downloaded from the backend server."""
    def __init__(self, message):
        super().__init__(message)
```

radmiratbrighterai replied:
added it

pkern90 commented on redact/v3/redact_requests.py:
```diff
@@ -218,6 +220,8 @@ def _stream_output_to_file(
                 else:
                     target_file.unlink(missing_ok=True)
 
+                    raise RuntimeError(f"failed to download the file {file}")
```
same as my comment for v4

radmiratbrighterai replied:
done

pkern90 reviewed (changes requested):
Please add tests. Make sure to adjust the return type hints from the changes in v3 from last PR. If I'm not mistaken they shouldn't be optional anymore.

radmiratbrighterai commented:
> Please add tests. Make sure to adjust the return type hints from the changes in v3 from last PR. If I'm not mistaken they shouldn't be optional anymore.

Added tests, removed the `Optional` return types.
